CMakeLists.txt

```diff
@@ -44,6 +47,14 @@ else()
     list(APPEND LIBXMP_SRC_LIST ${LIBXMP_SRC_LIST_PROWIZARD})
 endif()
 
+#lite-only defs
+list(APPEND LIBXMPLITE_DEFINES_PUBLIC -DLIBXMP_CORE_PLAYER)
+if(LIBXMP_DISABLE_IT)
+    list(APPEND LIBXMPLITE_DEFINES_PUBLIC -DLIBXMP_CORE_DISABLE_IT)
+endif()
```

Comment:
These defines are exported to the cmake files, but are not present in the libxmp headers.
Shouldn't these be private instead?

Reply:
I honestly don't know. I just followed the existing `LIBXMP_DEFINES_PUBLIC` in there. Maybe that's wrong too?

Reply:
I think so: looks like these are implementation-only defines.
You need to rename the variable to `LIBXMP(LITE)_DEFINES_PRIVATE` and add them as `PRIVATE` compile definitions.
But make sure to keep adding `LIBXMP_STATIC` as `PUBLIC`.
Also, you don't need to prepend the defines with `-D`, when used with `target_compile_definitions`.

Reply:
OK. I will force-push with the following applied: can you confirm? [cmake.patch.txt](https://github.com/libxmp/libxmp/files/15181914/cmake.patch.txt)

<details>

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 23c7bbe..1d3d7c3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,9 +2,8 @@ cmake_minimum_required(VERSION 3.2...3.5)
 project(libxmp VERSION 4.6.0 LANGUAGES C)
 
 set(LIBXMP_DEFINES)
-set(LIBXMP_DEFINES_PUBLIC)
+set(LIBXMPLITE_DEFINES)
 set(LIBXMP_CFLAGS)
-set(LIBXMPLITE_DEFINES_PUBLIC)
 
 include(${CMAKE_CURRENT_SOURCE_DIR}/cmake/libxmp-checks.cmake)
 include(GNUInstallDirs)
@@ -36,28 +35,28 @@ option(LIBXMP_DISABLE_IT            "Disable IT format in libXMP-lite" OFF)
 include(${CMAKE_CURRENT_SOURCE_DIR}/cmake/libxmp-sources.cmake)
 
 if(LIBXMP_DISABLE_DEPACKERS)
-    list(APPEND LIBXMP_DEFINES_PUBLIC -DLIBXMP_NO_DEPACKERS)
+    list(APPEND LIBXMP_DEFINES LIBXMP_NO_DEPACKERS)
 else()
     list(APPEND LIBXMP_SRC_LIST ${LIBXMP_SRC_LIST_DEPACKERS})
 endif()
 
 if(LIBXMP_DISABLE_PROWIZARD)
-    list(APPEND LIBXMP_DEFINES_PUBLIC -DLIBXMP_NO_PROWIZARD)
+    list(APPEND LIBXMP_DEFINES LIBXMP_NO_PROWIZARD)
 else()
     list(APPEND LIBXMP_SRC_LIST ${LIBXMP_SRC_LIST_PROWIZARD})
 endif()
 
 #lite-only defs
-list(APPEND LIBXMPLITE_DEFINES_PUBLIC -DLIBXMP_CORE_PLAYER)
+list(APPEND LIBXMPLITE_DEFINES LIBXMP_CORE_PLAYER)
 if(LIBXMP_DISABLE_IT)
-    list(APPEND LIBXMPLITE_DEFINES_PUBLIC -DLIBXMP_CORE_DISABLE_IT)
+    list(APPEND LIBXMPLITE_DEFINES LIBXMP_CORE_DISABLE_IT)
 endif()
 #for libxmp-lite.pc
 set(libxmplite_VERSION ${libxmp_VERSION})
 
 if(MSVC)
-    set(LIBXMP_MSVC_DEFINES -D_USE_MATH_DEFINES)
-    list(APPEND LIBXMP_DEFINES ${LIBXMP_MSVC_DEFINES})
+    list(APPEND LIBXMP_DEFINES _USE_MATH_DEFINES)
+    list(APPEND LIBXMPLITE_DEFINES _USE_MATH_DEFINES)
 endif()
 
 
@@ -76,7 +75,7 @@ if(BUILD_STATIC)
 
     target_compile_options(xmp_static PRIVATE ${LIBXMP_CFLAGS})
     target_compile_definitions(xmp_static PRIVATE ${LIBXMP_DEFINES})
-    target_compile_definitions(xmp_static PUBLIC ${LIBXMP_DEFINES_PUBLIC} -DLIBXMP_STATIC)
+    target_compile_definitions(xmp_static PUBLIC LIBXMP_STATIC)
 
     if(LIBXMP_PIC)
         target_compile_options(xmp_static PRIVATE -fPIC)
@@ -101,8 +100,8 @@ if(BUILD_STATIC)
         endif()
 
         target_compile_options(xmp_lite_static PRIVATE ${LIBXMP_CFLAGS})
-        target_compile_definitions(xmp_lite_static PRIVATE ${LIBXMP_DEFINES})
-        target_compile_definitions(xmp_lite_static PUBLIC ${LIBXMPLITE_DEFINES_PUBLIC} -DLIBXMP_STATIC)
+        target_compile_definitions(xmp_lite_static PRIVATE ${LIBXMPLITE_DEFINES})
+        target_compile_definitions(xmp_lite_static PUBLIC LIBXMP_STATIC)
 
         if(LIBXMP_PIC)
             target_compile_options(xmp_lite_static PRIVATE -fPIC)
@@ -157,11 +156,10 @@ if(BUILD_SHARED)
 
     target_compile_definitions(xmp_shared PRIVATE ${LIBXMP_DEFINES})
     if(WIN32)
-        target_compile_definitions(xmp_shared PRIVATE -DBUILDING_DLL)
+        target_compile_definitions(xmp_shared PRIVATE BUILDING_DLL)
     endif()
 
     target_compile_options(xmp_shared PRIVATE ${LIBXMP_CFLAGS})
-    target_compile_definitions(xmp_shared PUBLIC ${LIBXMP_DEFINES_PUBLIC})
 
     target_include_directories(xmp_shared PUBLIC "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>")
     target_include_directories(xmp_shared PUBLIC "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>")
@@ -211,13 +209,12 @@ if(BUILD_SHARED)
             message(STATUS "Versioned symbols: DISABLED")
         endif()
 
-        target_compile_definitions(xmp_lite_shared PRIVATE ${LIBXMP_DEFINES})
+        target_compile_definitions(xmp_lite_shared PRIVATE ${LIBXMPLITE_DEFINES})
         if(WIN32)
-            target_compile_definitions(xmp_lite_shared PRIVATE -DBUILDING_DLL)
+            target_compile_definitions(xmp_lite_shared PRIVATE BUILDING_DLL)
         endif()
 
         target_compile_options(xmp_lite_shared PRIVATE ${LIBXMP_CFLAGS})
-        target_compile_definitions(xmp_lite_shared PUBLIC ${LIBXMPLITE_DEFINES_PUBLIC})
 
         target_include_directories(xmp_lite_shared PUBLIC "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>")
         target_include_directories(xmp_lite_shared PUBLIC "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/libxmp-lite>")
```

Reply:
Looks good.

Before you make the change, these defines being public might be useful to users to know whether a feature is avaialble.
Perhaps that's why they were added to the public interface?

Reply:
> Looks good.

OK (but waiting for the question you raised)

> Before you make the change, these defines being public might be useful to users to know whether a feature is avaialble. Perhaps that's why they were added to the public interface?

Don't know, _possibly_.. Maybe @Wohlstand can remember? They have been there since the first addition of cmake support from #415 (but they dos't look that very much useful to me...)

Reply:
OK, force-pushed with cmake clean-ups.

Reply:
How do the tests know whether a feature is enabled and to be tested?

Reply:
test-dev, unfortunately, expects a full-feature build of the library.

Reply:
sigh.. test/test.c needs to know though: https://github.com/libxmp/libxmp/pull/725

Reply:
An alternative solution is to create an interface library, to which you add the various compile definitions.
e.g.
```cmake
add_library(libxmp-build-options INTERFACE)
if(LIBXMP_DISABLE_DEPACKERS)
  target_compile_definitions(libxmp-build-options INTERFACE LIBXMP_NO_DEPACKERS)
endif()

target_link_libraries(libxmp_static PRIVATE $<BUILD_INTERFACE:libxmp-build-options>)
target_link_libraries(libxmp_shared PRIVATE $<BUILD_INTERFACE:libxmp-build-options>)
target_link_libraries(libxmp_lite_static PRIVATE $<BUILD_INTERFACE:libxmp-build-options>)
target_link_libraries(libxmp_lite_shared PRIVATE $<BUILD_INTERFACE:libxmp-build-options>)
target_link_libraries(libxmp-test PRIVATE $<BUILD_INTERFACE:libxmp-build-options>)
```

Reply:
Will try later, thanks. (Patches are welcome, btw :)